src/Microsoft.AspNetCore.Routing/Tree/TreeRouter.cs

```diff
@@ -275,19 +273,23 @@ public bool MoveNext()
                         Current = next;
                         return true;
                     }
-                    else if (++_segmentIndex >= _tokenizer.Count)
+                    // Next template has the same length as the url we are trying to match
+                    // The only possible matching segments are either our current matches or
+                    // any catch-all segment after this segment in which the catch all is empty.
+                    else if (next.Length == _tokenizer.Count)
                     {
-                        _segmentIndex--;
                         if (next.Matches.Count > 0)
                         {
                             Current = next;
                             return true;
                         }
-                    }
-
-                    if (_tokenizer.Count == 0)
```

Comment:
This case is not required as it's covered by https://github.com/aspnet/Routing/pull/317/files#diff-dc035bd90601467ff505d414a73d85deR279